app/cdap/components/RecipeManagement/CreateRecipe/index.tsx

```diff
@@ -0,0 +1,182 @@
+  };
+
+  // This static data has to be removed when we have actual API data, then directly we will get that data from store as directives
+  const recipeSteps = ['uppercase: body1', 'titlecase: body2'];
+
+  const handleSaveButtonMode = (
+    formData: IRecipeFormData = recipeFormData,
+    nameErrorData: IRecipeNameErrorData = recipeNameErrorData
+  ) => {
+    if (
+      formData.recipeName === '' ||
+      formData.description === '' ||
+      formData.recipeName?.trim().length === 0 ||
+      formData.description?.trim().length === 0 ||
+      nameErrorData.isRecipeNameError
+    ) {
+      setIsSaveDisabled(true);
+    } else {
+      setIsSaveDisabled(false);
+    }
+  };
+
+  const handleRecipeFormData = (formData: IRecipeFormData) => {
+    setRecipeFormData(formData);
+    handleSaveButtonMode(formData);
+  };
+
+  const onRecipeNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    handleRecipeFormData({
+      ...recipeFormData,
+      recipeName: event.target.value,
+    });
+    validateRecipeNameExists.current({
+      recipeName: event.target.value,
+      description: recipeFormData.description,
+    });
+  };
+
+  const onRecipeDescriptionChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    handleRecipeFormData({
+      ...recipeFormData,
+      description: event.target.value,
+    });
+  };
+
+  const onFormSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    onRecipeDataSave(recipeFormData);
+  };
+
+  const onRecipeDataSave = (recipeFormData: IRecipeFormData) => {
+    const requestBody = {
+      recipeName: recipeFormData.recipeName,
+      description: recipeFormData.description,
+      directives: recipeSteps,
+    };
+    createRecipeService({ requestBody, onCreateRecipeResponse, onCreateRecipeError });
+  };
+
+  const onCreateRecipeResponse = () => {
+    setShowRecipeForm(false);
+    setSnackbar({
+      open: true,
+      isSuccess: true,
+      message: `${recipeSteps.length} ${T.translate(`${PREFIX}.recipeSaveSuccessMessage`)}`,
+    });
+  };
+
+  const onCreateRecipeError = (err: Record<string, unknown>) => {
+    setShowRecipeForm(false);
+    setSnackbar({
+      open: true,
+      isSuccess: false,
+      message: (err.response as Record<string, string>).message,
+    });
+  };
+
+  const onCancel = () => {
+    setShowRecipeForm(false);
+  };
+
+  const onGetRecipeByNameError = (err: Record<string, unknown>, formData: IRecipeFormData) => {
+    if (err.statusCode === 404) {
+      setRecipeNameErrorData(noErrorState, formData);
+    }
+  };
+
+  const onGetRecipeByNameResponse = (formData: IRecipeFormData) => {
+    !recipeNameErrorData.isRecipeNameError &&
+      setRecipeNameErrorData(
+        {
+          isRecipeNameError: true,
+          recipeNameErrorMessage: T.translate(`${PREFIX}.sameNameErrorMessage`).toString(),
+        },
+        formData
+      );
+  };
+
+  const validateRecipeNameExists = useRef(
```

Comment:
In this function we are validating recipe name input filed (whether recipe name already exists or not and recipe name without alphanumeric characters)  based on the result we are showing the helper text

Reply:
Thanks for adding the comment - please split it over multiple lines, since it's long.

In this flow, it looks like there are several set state calls in a row. You could consider using `useReducer` to consolidate the logic and avoid race conditions.

Reply:
used reducer instead of useState for setRecipeNameErrorDataState which is used in this function

Reply:
The reducer only sets a single field? That's not an effective use for a reducer, especially since the logic for disabling the save button is still in the component.

Reply:
used State everywhere instead of reducer